src/lsp/cobol_parser/expect/gen_expect.ml

```diff
@@ -0,0 +1,421 @@
+
+let emit_complentryset ppf custom_types =
+  Fmt.(pf ppf {|
+module CompEntrySet = Set.Make(struct
+  type t = completion_entry
+  let compare ce1 ce2 =
+    let to_int = function
+      %a| K _ -> %d in
+    match ce1, ce2 with
+      | K nel1, K nel2 -> NEL.compare Stdlib.compare nel2 nel1
+      | _ -> to_int ce1 - to_int ce2
+end)
+|}
+  (list ~sep:nop (fun ppf (i,t) ->
+    pf ppf "| %s -> %d\n    " t i)) (List.mapi (fun i v -> (i,v)) custom_types)
+  (List.length custom_types))
+
+let emit_completion_entry ppf =
+  Fmt.pf ppf "type completion_entry =\n";
+  Fmt.pf ppf "  | K of token NEL.t\n";
+  let custom_types = Nonterminal.fold (fun nonterm acc ->
+    match nonterminal_filter_map nonterm with
+    | Some (Custom s) -> (Fmt.pf ppf "  | %s\n" s; s::acc)
+    | _ -> acc) [] in
+  emit_pp_completion_entry ppf custom_types;
+  emit_complentryset ppf custom_types
+
+let nullable_nonterminals = Lr1.tabulate begin fun lr1 ->
+  SymbolSet.from_list_map ~add:SymbolSet.add_option
+  (fun (prod, idx) ->
+    match (Production.rhs prod).(idx) with
+      | N nt, _, _ when Nonterminal.nullable nt -> Some (N nt)
+      | _ | exception Invalid_argument _ -> None)
+    (Lr0.items @@ Lr1.lr0 lr1)
+  end
+
+let reducable_productions = Lr1.tabulate begin fun lr1 ->
+  List.filter_map (function
+    | (prod, idx) when Array.length (Production.rhs prod) == idx ->
+        Some prod
+    | _ -> None)
+  (Lr0.items @@ Lr1.lr0 lr1)
+  |> List.sort_uniq Production.compare
+end
+
+let emit_accaptable_nullable_nonterminals_in_env ppf =
+  Fmt.pf ppf {|
+let accaptable_nullable_nonterminals_in ~env : Menhir.xsymbol list =
+  Menhir.(match current_state_number env with
+|};
+  Lr1.fold (fun lr1 acc ->
+    match nullable_nonterminals lr1 with
+    | emtpy when SymbolSet.is_empty emtpy -> acc
+    | symbols -> SymbolMap.add_to_list symbols lr1 acc)
+  SymbolMap.empty
+  |> inv (SymbolMap.fold (fun s l acc -> (l,
+      List.map (function  N nt -> nt | T _ -> failwith "impossible" ) @@ SymbolSet.elements s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    (pp_brackets @@ pp_list @@ pp_xsymbol_of_nt)
+    "[])"
+
+let emit_reducable_productions_in_env ppf =
+  Fmt.pf ppf {|
+let reducable_productions_in ~env : Menhir.production list =
+  List.rev_map Menhir.find_production @@@@
+  match Menhir.current_state_number env with
+|};
+  Lr1.fold (fun lr1 acc ->
+    match reducable_productions lr1 with
+    | [] -> acc
+    | defaults -> ProdMap.add_to_list defaults lr1 acc)
+  ProdMap.empty
+  |> inv (ProdMap.fold (fun s l acc -> (l, s)::acc)) []
+  |> pp_match_cases ppf
+    Fmt.(using Lr1.to_int int)
+    Fmt.(pp_brackets @@ pp_list @@ (using Production.to_int int))
+    "[]"
+
+let completion_entries_of =
+  let rec continue_while_terminal rhs i l =
+    match rhs.(i) with
+        | T t, _, _ when is_valid_for_comp t ->
+            continue_while_terminal rhs (i+1) (t::l)
+        | _ | exception Invalid_argument _ -> List.rev l
```

Comment:
In general, if at least one case expression starts on its own line, please break lines right after `->` (and before any `when`), even if that leads to short lines.